pkg/controller/resourceclaim/controller.go

```diff
@@ -854,12 +857,52 @@ func (ec *Controller) syncClaim(ctx context.Context, namespace, name string) err
 		// those, the resource claim controller will trigger deletion when the
 		// pod is done. However, it doesn't hurt to also trigger deallocation
 		// for such claims and not checking for them keeps this code simpler.
-		if len(valid) == 0 &&
-			claim.Spec.AllocationMode == resourcev1alpha2.AllocationModeWaitForFirstConsumer {
-			claim.Status.DeallocationRequested = true
+		if len(valid) == 0 {
+			if builtinControllerFinalizer >= 0 {
+				if claim.Spec.AllocationMode == resourcev1alpha2.AllocationModeWaitForFirstConsumer ||
+					claim.DeletionTimestamp != nil {
+					// Allocated by scheduler with structured parameters. We can "deallocate"
+					// by clearing the allocation.
+					claim.Status.Allocation = nil
+				}
+			} else if claim.Spec.AllocationMode == resourcev1alpha2.AllocationModeWaitForFirstConsumer {
+				// DRA driver controller in the control plane
+				// needs to do the deallocation.
+				claim.Status.DeallocationRequested = true
+			}
+			// In all other cases, we keep the claim allocated, in particular for immediate allocation
+			// with a control plane controller.
+		}
+
+		claim, err := ec.kubeClient.ResourceV1alpha2().ResourceClaims(claim.Namespace).UpdateStatus(ctx, claim, metav1.UpdateOptions{})
+		if err != nil {
+			return err
 		}
 
-		_, err := ec.kubeClient.ResourceV1alpha2().ResourceClaims(claim.Namespace).UpdateStatus(ctx, claim, metav1.UpdateOptions{})
+		// Now also remove the finalizer. Note that the index may have changed.
+		builtinControllerFinalizer := slices.Index(claim.Finalizers, resourcev1alpha2.Finalizer)
+		if builtinControllerFinalizer >= 0 {
+			claim.Finalizers = slices.Delete(claim.Finalizers, builtinControllerFinalizer, builtinControllerFinalizer+1)
+			if _, err := ec.kubeClient.ResourceV1alpha2().ResourceClaims(claim.Namespace).Update(ctx, claim, metav1.UpdateOptions{}); err != nil {
+				return err
+			}
+		}
+	} else if builtinControllerFinalizer >= 0 && claim.DeletionTimestamp != nil && len(valid) == 0 {
+		claim := claim.DeepCopy()
+		if claim.Status.Allocation != nil {
+			// This can happen when a claim with immediate allocation
+			// stopped being used, remained allocated, and then got
+			// deleted. As above we then need to clear the allocation.
+			claim.Status.Allocation = nil
+			var err error
+			claim, err = ec.kubeClient.ResourceV1alpha2().ResourceClaims(claim.Namespace).UpdateStatus(ctx, claim, metav1.UpdateOptions{})
+			if err != nil {
+				return err
+			}
+		}
+		// Whether it was allocated or not, remove the finalizer to unblock removal.
+		claim.Finalizers = slices.Delete(claim.Finalizers, builtinControllerFinalizer, builtinControllerFinalizer+1)
+		_, err := ec.kubeClient.ResourceV1alpha2().ResourceClaims(claim.Namespace).Update(ctx, claim, metav1.UpdateOptions{})
```

Comment:
These 2 updates seems a bit excessive. Why not just do a single update once here? What's the benefit of doing status update a few lines above.

Reply:
That's what I was doing earlier until it was pointed out that the status update shouldn't be allowed to touch finalizers. The type implementation (like many others) was buggy in that regard. I fixed that, but now we have to do two calls, one for the object (here) and one for the status (above).

Reply:
https://github.com/kubernetes/kubernetes/pull/123516#discussion_r1513276787

Reply:
I guess that should trigger a discussion about whether the allocation should be part of the spec instead?
For reference, the nodeName is part of a PodSpec, not the status.

Maybe there should be a dedicated endpoint just for it? Analogous to Pod binding. And we could validate in the strategy that the allocation is not updated via the regular endpoint.

But it's a discussion for 1.31.